### ISSUE TITLE
Webview on transparent WPF regions does not hittest.

### ISSUE THREAD

**Issue by Skarnivorous** [MicrosoftEdge/WebView2Feedback#997]
**Description**
When WPF window has Allow Transparency and no style (completely transparent window).
And WebView2 is on window region where no other controls are visible (i.e. Panel hosting WebView2 are background transparent / opacity 0).
And WebView2 content is visible (though it may have opacity).
Clicks on WebView2 will pass through WebView2, not trigger webview2 content events, and activate desktop UI beneath the window.

Expectation would be that the WPF hit test takes into account WebView2's visible painted areas.
Ideal would also treat non-painted areas as pass-through.

**Version**
SDK: 1.0.790-prerelease
Runtime: Version 90.0.809.0
Framework: WPF, C#, XAML
OS: Windows 10 Pro Version 20H2 OS build 19042.804 Windows Feature Experience Pack 120.2212.551.0

**Additional context**
With the new DefaultBackgroundColor allowing transparency I have successfully rendered WebView2 content in a transparent window with alpha blending/composition on the desktop. The bug above however means that there can be no interaction with the content as the entire control is not hittested even for the areas of the control which are visible.

Re-enabling window background, or having other wpf controls visible behind or in-front of the webview2 will suddenly make the webview2 receive hittest in the regions where the other controls overlap.

**Comment by darbid** [comment]
Can you had an image of what you are talking about? Unfortunately your text is indicating something that is impossible for the webview2 as it is not a true wpf control but a wrapped winforms control and thus allows transparency in wpf will effect the wibview2

**Comment by Skarnivorous** [comment]
I have linked a screen video to demonstrate. In the video the red box is the webview2 and the clock is a normal image control. You can see that when I click a region in the webview2 that overlaps the clock, the webview2 will receive the click event. However when I click a region in the webview2 that is only overlaying the desktop, the click will pass straight through. I believe the webview2 should still receive the hittest / mouse event as it is still visible. All WPF layer/visual tree type calls also ignore the webview2 in those areas. I would expect if the webview2 has 100% transparency in some region, then it should have this correct behaviour of click passthrough.

https://streamable.com/szltec

**Comment by Skarnivorous** [comment]
I am hoping this can be realised as possible via native implementation as I have succeeded using a workaround using CapturePreviewAsync running continuously but I imagine this is a slow routine.

**Comment by pagoe-msft** [comment]
@Skarnivorous 

Thanks for submitting this - I've added this item to our backlog.

**Comment by Blightbuster** [comment]
> I am hoping this can be realised as possible via native implementation as I have succeeded using a workaround using CapturePreviewAsync running continuously but I imagine this is a slow routine.

Would you mind sharing your workaround implementation? I assume you overwrote the HitTestCore method of the WPF WebView2 contol?

**Comment by Skarnivorous** [comment]
No, I took snapshots of the webview2 control content (which preserved
transparency) and then used that as an image control source under the
webview2. This made the image control the same hittest "shape" as the
webview2 content so the webview2 then received the unser interaction only
where there was visible content. It worked but the sample rate of the
snapshots was so slow, the image control content and the webview2 content
was out of sync by about 200-1000ms. I was working with animated webview2
content so this was not enough. If your webview2 content is fairly static,
this may work for you.

**Comment by xu-ms** [comment]
@Skarnivorous 
I try to reproduce your scenario, I've also created an image and WebView control and set `webView2Control.DefaultBackgroundColor = System.Drawing.Color.Transparent`. It seems that when I click on the transparent area, it is still responded to by the WebView, and the text can be selected. Perhaps my understanding is incorrect; can you please explain further?
![Animation](https://user-images.githubusercontent.com/118523543/229052252-a4d314e4-55d6-4c92-b6f1-db790e66416b.gif)

**Comment by Skarnivorous** [comment]
Thanks for investigating! - the feedback / issue is that the usual
behaviour of WPF/browser controls in this scenario is expected to pass the
hit test / click through the transparent areas to the parent controls >
parent app window > desktop app windows or desktop below the web view
control.

I understand it may be the current behaviour of the control - but that it
would be an improved behaviour to work in this manner, perhaps optionally.

**Comment by menghuijinxi** [comment]
我找到了一个解决该问题的方法，但它不是通用方案，适用场景有限，可能无法满足所有需求。相比传统的基于截图和像素比对的输入检测方式，该方案在性能上有明显提升。

实现思路如下：

整个网页作为 UI 覆盖层完全覆盖在后端应用（如 Unreal Engine）窗口之上。然而，网页中很大一部分区域并不需要前端交互（例如背景区域或仅用于展示的 UI），这些区域的输入应由后端直接处理。

我们可以通过 JavaScript 监听键盘事件，并判断当前是否需要前端响应。如果当前上下文不需要前端处理输入（比如光标位于非交互区域），则通过 window.chrome.webview.postMessage() 将按键事件转发给后端（也可使用 WebSocket 等其他通信方式）。

后端接收到消息后，通过模拟原生平台输入事件的方式触发对应行为。例如，在 Unreal Engine 中，我通过调用底层平台消息处理接口：

`FSlateApplication::Get().GetPlatformApplication()->GetMessageHandler()->OnKeyDown(...)`

来模拟向平台发送键盘按下消息。这种方式绕过了前端 UI 系统，直接注入输入事件，从而实现“输入穿透”的效果 —— 即：网页虽覆盖窗口，但特定区域的输入仍能被后端正确接收和响应。

**Comment by Paper-Folding** [comment]
> 我找到了一个解决该问题的方法，但它不是通用方案，适用场景有限，可能无法满足所有需求。相比传统的基于截图和像素比对的输入检测方式，该方案在性能上有明显提升。
> 
> 实现思路如下：
> 
> 整个网页作为 UI 覆盖层完全覆盖在后端应用（如 Unreal Engine）窗口之上。然而，网页中很大一部分区域并不需要前端交互（例如背景区域或仅用于展示的 UI），这些区域的输入应由后端直接处理。
> 
> 我们可以通过 JavaScript 监听键盘事件，并判断当前是否需要前端响应。如果当前上下文不需要前端处理输入（比如光标位于非交互区域），则通过 window.chrome.webview.postMessage() 将按键事件转发给后端（也可使用 WebSocket 等其他通信方式）。
> 
> 后端接收到消息后，通过模拟原生平台输入事件的方式触发对应行为。例如，在 Unreal Engine 中，我通过调用底层平台消息处理接口：
> 
> `FSlateApplication::Get().GetPlatformApplication()->GetMessageHandler()->OnKeyDown(...)`
> 
> 来模拟向平台发送键盘按下消息。这种方式绕过了前端 UI 系统，直接注入输入事件，从而实现“输入穿透”的效果 —— 即：网页虽覆盖窗口，但特定区域的输入仍能被后端正确接收和响应。

这个issue是要求webview2窗口的透明背景鼠标可直接点击透过，能够点击到窗口下面的其他应用/桌面的元素，不是说使得部分区域无鼠标交互。

**Comment by Paper-Folding** [comment]
When a transparent window contains webview2, and webview2 also have transparent background, if hit test can bypass transparent area, that will be a super helpful feature.  
On a native WPF window with transparent background, this feature is implemented by default.
